File: apps/backend-node/src/features/bookings/booking-pricing.service.ts
import { Injectable } from '@nestjs/common';

import { PrismaService } from '../../prisma/prisma.service';
import { BadRequestError } from '../../shared/errors';

const PLATFORM_SETTINGS_ID = 1;

export interface BookingPrice {
  rentKrw: number;
  serviceFeeKrw: number;
  totalKrw: number;
  pricingVersion: string;
}

@Injectable()
export class BookingPricingService {
  constructor(private readonly prisma: PrismaService) {}

  async calculatePrice(
    monthlyPriceKrw: number,
    nights: number,
  ): Promise<BookingPrice> {
    if (monthlyPriceKrw <= 0) {
      throw new BadRequestError('Monthly price must be positive');
    }

    if (nights <= 0) {
      throw new BadRequestError('Stay length must be positive');
    }

    const settings = await this.getPlatformSettings();
    const rentKrw = roundDownToThousands((monthlyPriceKrw / 30) * nights);
    const serviceFeeKrw = roundDownToThousands(
      rentKrw * (settings.service_fee_bps / 10_000),
    );

    return {
      rentKrw,
      serviceFeeKrw,
      totalKrw: rentKrw + serviceFeeKrw,
      pricingVersion: settings.pricing_version,
    };
  }

  async getHoldTtlMinutes(): Promise<number> {
    const settings = await this.getPlatformSettings();
    return settings.hold_ttl_minutes;
  }

  private async getPlatformSettings() {
    const settings = await this.prisma.platform_settings.findUnique({
      where: { id: PLATFORM_SETTINGS_ID },
    });

    if (!settings) {
      throw new BadRequestError('Platform settings are not configured');
    }

    return settings;
  }
}

function roundDownToThousands(value: number): number {
  return Math.floor(value / 1000) * 1000;
}

File: apps/backend-java/src/main/java/com/housingplatform/features/bookings/BookingPricingService.java
package com.housingplatform.features.bookings;

import com.housingplatform.shared.auth.error.BadRequestException;
import com.housingplatform.features.bookings.model.BookingPrice;
import com.housingplatform.persistence.entity.PlatformSetting;
import com.housingplatform.persistence.repository.PlatformSettingRepository;
import org.springframework.stereotype.Service;

@Service
public class BookingPricingService {

  private static final short PLATFORM_SETTINGS_ID = 1;

  private final PlatformSettingRepository platformSettingRepository;

  public BookingPricingService(PlatformSettingRepository platformSettingRepository) {
    this.platformSettingRepository = platformSettingRepository;
  }

  public BookingPrice calculatePrice(int monthlyPriceKrw, int nights) {
    if (monthlyPriceKrw <= 0) {
      throw new BadRequestException("Monthly price must be positive");
    }
    if (nights <= 0) {
      throw new BadRequestException("Stay length must be positive");
    }

    PlatformSetting settings =
        platformSettingRepository
            .findById(PLATFORM_SETTINGS_ID)
            .orElseThrow(() -> new BadRequestException("Platform settings are not configured"));

    int rentKrw = roundDownToThousands((monthlyPriceKrw / 30.0) * nights);
    int serviceFeeKrw =
        roundDownToThousands(rentKrw * settings.getServiceFeeBps() / 10_000.0);
    int totalKrw = rentKrw + serviceFeeKrw;

    return new BookingPrice(rentKrw, serviceFeeKrw, totalKrw, settings.getPricingVersion());
  }

  public int getHoldTtlMinutes() {
    return platformSettingRepository
        .findById(PLATFORM_SETTINGS_ID)
        .map(PlatformSetting::getHoldTtlMinutes)
        .orElseThrow(() -> new BadRequestException("Platform settings are not configured"));
  }

  private static int roundDownToThousands(double value) {
    return (int) (Math.floor(value / 1000.0) * 1000);
  }
}
